pkg/proc/stack.go

```diff
@@ -93,19 +93,29 @@ func (frame *Stackframe) FramePointerOffset() int64 {
 // ThreadStacktrace returns the stack trace for thread.
 // Note the locations in the array are return addresses not call addresses.
 func ThreadStacktrace(thread Thread, depth int) ([]Stackframe, error) {
+	frames, _, err := ThreadStackframes(thread, 0, depth)
+	return frames, err
+}
+
+// ThreadStackframes returns the stack trace for thread.
+// Note the locations in the array are return addresses not call addresses.
+// ThreadStackframes also returns the next index to be used as 'start' argument
+// while scanning for all available stack frames, or -1 if there was an error
+// or if the index already reached the last possible value.
+func ThreadStackframes(thread Thread, start, depth int) ([]Stackframe, int, error) {
```

Comment:
I don't like the Stacktrace/Stackframes split where one is a wrapper for calling the other. I think there should be only one of each pair. But also, why are you doing the thing with `next` instead of just keeping the iterator?

Reply:
IMHO this is introducing a lot of machinery, both in service/dap and in stack.go for what seems like a meager optimization. You could call `Stacktrace(goroutineID, request.Arguments.StartFrame + request.Arguments.Levels, 0)` and get the same effect. Since the stacktrace always starts from the top you are saving very little work (just loading inlined frames, of which there will be none in non-optimized code) and some memory (not saving frames before the start).

I think there are three better solutions possible:

1. Add a `skip` member to stackIterator, keep `proc.(*stackIterator).stacktrace` and `proc.(*stackIterator).appendInlinedCalls` identical except for the `append`s, when `it.skip > 0` instead of `append`ing it will decrement `it.skip`. This removes the nextStackFrame map from service/dap and the weird new loop in `(*stackIterator).stacktrace`, the `Stackframe/Stacktrace` split and the new return value they would need. You lose a small amount of work saved (which will be nothing anyway on optimized code) but keep the memory save and the `proc` functions have a reasonable APIs.
2. Keep the stackIterator. This gets rid of the `Stackframe/Stacktrace` split and you actually make stack loading not be O(n**2) (although, it probably doesn't matter given that it's driven by user input and will produce unreasonably long stacks long before it becomes slow)
3. Do not care about the performance and just call `Stacktrace(StartFrame+Levels)` always, maybe add a limit to `Levels - StartFrame` or something.

BTW you don't need to ask for more frames to determine if you have reached the bottom of the stack, if the last frame is an error or has `Bottom` set, then it's the bottom frame. You can just send `len(frames)+arbitraryNumber` to the client if the last frame isn't the bottom of the stack.

Reply:
I made the change to load all of the frames up to start + levels and added the check if Bottom is set.